impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java

```diff
@@ -374,22 +368,27 @@ public synchronized void close() throws Exception {
      * Workaround for <a href="https://issues.apache.org/jira/browse/KAFKA-14287">KAFKA-14287</a>.
      * with kraft, if we don't shut down the controller last, we sometimes see a hang.
      */
-    private void roleOrderedShutdown(Map<Integer, Server> servers, boolean await) {
+    private void roleOrderedShutdown(Map<Integer, Server> servers, boolean await, Consumer<Integer> onBrokerStopped) {
         var justBrokers = servers.entrySet().stream()
                 .filter(e -> !portsAllocator.hasRegisteredPort(Listener.CONTROLLER, e.getKey()))
-                .map(Map.Entry::getValue)
-                .toList();
-        justBrokers.forEach(Server::shutdown);
-        if (await) {
-            justBrokers.forEach(Server::awaitShutdown);
-        }
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        shutdownNodes(justBrokers, await, onBrokerStopped);
 
         var controllers = servers.entrySet().stream()
                 .filter(e -> portsAllocator.hasRegisteredPort(Listener.CONTROLLER, e.getKey()))
-                .map(Map.Entry::getValue).toList();
-        controllers.forEach(Server::shutdown);
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        shutdownNodes(controllers, await, onBrokerStopped);
+    }
+
+    private static void shutdownNodes(Map<Integer, Server> nodeMap, boolean await, Consumer<Integer> onBrokerStopped) {
+        nodeMap.forEach((brokerId, server) -> server.shutdown());
         if (await) {
-            controllers.forEach(Server::awaitShutdown);
+            nodeMap.forEach((brokerId, server) -> {
+                server.awaitShutdown();
+                if (onBrokerStopped != null) {
+                    onBrokerStopped.accept(brokerId);
```

Comment:
It seems odd that the onBrokerStopped doesn't get call at all in the await false case.

Reply:
Yeah but I took the view that we didn't know it was stopped so didn't invoke the callback. 

In hindsight that's probably not helpful it should just invoke.